src/platform/hashtable.c

```diff
@@ -1389,11 +1402,82 @@ CxPlatHashTableContract(
 
         if (HT_SECOND_LEVEL_DIR_MIN_SIZE == HashTable->TableSize) {
             HashTable->SecondLevelDir = FirstLevelDir[0];
-            CXPLAT_FREE(FirstLevelDir);
+            CXPLAT_FREE(FirstLevelDir, QUIC_POOL_HASHTABLE);
         }
     }
 
     return TRUE;
 }
 
-#endif // CXPLAT_HASHTABLE_RESIZE_SUPPORT
+#endif // CXPLAT_HASHTABLE_CONTRACT_SUPPORT
+
+//
+// The maximum percentage of empty buckets in the hash table. If a hash table
+// has more empty buckets, it needs to be restructured.
+//
+#define CXPLAT_HASHTABLE_MAX_EMPTY_BUCKET_PERCENTAGE    25
+
+//
+// The maximum number of hash table restructions allowed at a time. This limits
+// the time spent expanding/contracting a hash table at dispatch.
+//
+#define CXPLAT_HASHTABLE_MAX_RESTRUCT_ATTEMPTS          8
```

Comment:
IMHO, this number seems a bit high. Maybe 2 or 4 instead?

Reply:
Yeah idk - it's a magic number. Probably even `1` would suffice as long as the load factor is greater than `1` since any single failure to expand will be compensated on the next hashtable insert.

Reply:
It's absolutely an edge case, though. I'd just leave it alone until data indicates otherwise.

Reply:
Since we're going to call this after every single insert, why not just 1 restructure at a time?

Reply:
Yeah, fine by me.